### DISCUSSION TITLE
Configuring OCSP FQDN Resolution on a specific network interface

### DISCUSSION THREAD

**Discussion by nknkgithub** [strongswan/strongswan#2257] [Q&A]
We are using strongswan 5.9.14 with OCSP enabled in strongswan.conf and revocation=ifuri set in swanctl.conf.
Currently, the OCSP FQDN resolution occurs on interface where default route is configured.
We need the OCSP FQDN resolution to be performed using a different interface(i.e netcsec). How can we achieve this?

Note: Currently netcsec interface is used to communicate with security gateway

Pasting the below outputs
ip addr show
```
ip a s
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
4: eth0@if431: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default
    link/ether da:1c:6c:b5:0f:a1 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 172.16.192.124/32 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::d81c:6cff:feb5:fa1/64 scope link
       valid_lft forever preferred_lft forever
9: neto1c: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 0a:d5:60:02:00:03 brd ff:ff:ff:ff:ff:ff
    altname enp1s16f7
    inet6 2005:11:54:97::1/64 scope global
       valid_lft forever preferred_lft forever
    inet6 fe80::8d5:60ff:fe02:3/64 scope link
       valid_lft forever preferred_lft forever
10: netcore: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 0a:d5:60:02:00:04 brd ff:ff:ff:ff:ff:ff
    altname enp1s17f1
    inet6 2008:1:161:19::2/128 scope global nodad
       valid_lft forever preferred_lft forever
    inet6 fe80::8d5:60ff:fe02:4/64 scope link
       valid_lft forever preferred_lft forever
17: netcsec: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 0a:d5:60:02:00:0b brd ff:ff:ff:ff:ff:ff
    inet 10.210.106.211/24 brd 10.210.106.255 scope global netcsec
       valid_lft forever preferred_lft forever
    inet6 fe80::8d5:60ff:fe02:b/64 scope link
       valid_lft forever preferred_lft forever
44: netf1c: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 0a:d5:60:02:00:26 brd ff:ff:ff:ff:ff:ff
    altname enp138s1
    inet6 2001:4000:aa:71::1/64 scope global
       valid_lft forever preferred_lft forever
    inet6 fe80::8d5:60ff:fe02:26/64 scope link
       valid_lft forever preferred_lft forever
59: nete1c: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 0a:d5:60:02:00:27 brd ff:ff:ff:ff:ff:ff
    altname enp138s1f1
    inet6 2001:4000:aa:73::1/64 scope global
       valid_lft forever preferred_lft forever
    inet6 fe80::8d5:60ff:fe02:27/64 scope link
       valid_lft forever preferred_lft forever
```
ip route show
```
ip route show
default via 169.254.1.1 dev eth0
169.254.1.1 dev eth0 scope link
```

ip rule list
```
ip rule list
0:      from all lookup local
220:    from all lookup 220
32765:  from 10.210.106.211 lookup 101
32766:  from all lookup main
32767:  from all lookup default
```

ip route show table 101
```
ip route show table 101
default via 10.210.106.254 dev netcsec
10.210.106.0/24 dev netcsec proto kernel scope link src 10.210.106.211
```

**Comment by tobiasbrunner** [comment]
> Currently, the OCSP FQDN resolution occurs on interface where default route is configured.
> We need the OCSP FQDN resolution to be performed using a different interface(i.e netcsec). How can we achieve this?

Configure the routes differently?

**Reply by nknkgithub** [reply]
Thanks @tobiasbrunner .

This routing table and rule list are from the pod.
The default route is configured by the calico CNI used in Kubernetes.
For tunnel establishment we use netcsec interface through the routing table 101 using source based rule.

```
ip rule list
0:      from all lookup local
220:    from all lookup 220
32765:  from 10.210.106.211 lookup 101
32766:  from all lookup main
32767:  from all lookup default
```
ip route show table 101
```
ip route show table 101
default via 10.210.106.254 dev netcsec
10.210.106.0/24 dev netcsec proto kernel scope link src 10.210.106.211
```

Does the charon daemon tries to resolve FQDN over all the interfaces if it is unable to resolve it via default interface?

**Reply by tobiasbrunner** [reply]
You mentioned OCSP before, that happens in an external library (usually _libcurl_), so you have to check there. But probably it will just boil down to a call to `getaddrinfo()` (like the daemon does for IKE) i.e. this queries the system's resolver(s). But the name resolution is not related to how the traffic to the resolved server are routed afterwards.

**Comment by nknkgithub** [comment]
Thanks @tobiasbrunner .

Does it mean that OCSP URL will be resolved using libcurl, but traffic towards resolved IP(OCSP responder) will be sent using source IP mentioned in "local_addrs=" parameter in swanctl.conf?

We tested this scenario. Observation is OCSP request is going via default interface. But we want OCSP request to go via IP mentioned in "local_addrs=" parameter. How this can be achieved?

**Reply by tobiasbrunner** [reply]
> Does it mean that OCSP URL will be resolved using libcurl, but traffic towards resolved IP(OCSP responder) will be sent using source IP mentioned in "local_addrs=" parameter in swanctl.conf?

No, the whole OCSP query is handled by the fetcher i.e. libcurl (it's a single request/response) and it has nothing to do with `local_addrs`.

> But we want OCSP request to go via IP mentioned in "local_addrs=" parameter.

Why?

> How this can be achieved?

Without code changes, the only way would be via routing changes or firewall rules. In the code, you could instruct libcurl to bind to a specific interface/IP via `CURLOPT_INTERFACE` if that's what you want.

**Comment by nknkgithub** [comment]
Thanks @tobiasbrunner .

OCSP request goes via default interface or via "local_addrs" configured swanct.conf?
